R/pmcmc_nested_parameters.R

```diff
@@ -0,0 +1,488 @@
+
+      if (!is.null(private$fixed_parameters)) {
+        lp_fix <- private$fixed_parameters$prior(
+          theta[1, private$param_names$fixed])
+      } else {
+        lp_fix <- 0
+      }
+
+      lp_vary <- vnapply(self$populations(),
+        function(x) private$varied_parameters[[x]]$prior(
+          theta[x, private$param_names$varied]))
+
+      lp_fix + lp_vary
+    },
+
+    ##' @description Propose a new parameter vector given a current parameter
+    ##' vector. This proposes a new parameter vector/matrix given your current
+    ##' vector and the variance-covariance matrix of your proposal
+    ##' kernel, discretises any discrete values, and reflects bounded
+    ##' parameters until they lie within `min`:`max`. Returns matrix if any
+    ##' varied parameters are included, otherwise vector.
+    ##'
+    ##' @param theta a parameter vector in the same order as your
+    ##' parameters were defined in (see `$names()` for that order.
+    ##'
+    ##' @param scale an optional scaling factor to apply to the
+    ##' proposal distribution. This may be useful in sampling starting
+    ##' points. The parameter is equivalent to a multiplicative factor
+    ##' applied to the variance covariance matrix.
+    ##'
+    ##' @param type specifies which type of parameters should be proposed,
+    ##' either fixed parameters only ("fixed"), varied only ("varied"), or
+    ##' both ("both") types. For 'fixed' and 'varied' then parameters of the
+    ##' other type is returned equal to corresponding value in `theta`.
+    propose = function(theta, scale = 1, type = c("fixed", "varied", "both")) {
+      type <- match.arg(type)
+      pfix <- function(theta, scale) {
+        if (!is.null(private$proposal_fixed)) {
+          private$proposal_fixed(theta, scale)
+        }
+      }
+      pvar <- function(theta, scale) {
+        var <- vapply(seq(nrow(theta)), function(i)
+          private$proposal_varied[[rownames(theta)[[i]]]](theta[i, ], scale),
+          numeric(length(self$names())))
+        # catch one varied edge case
+        if (!inherits(var, "matrix")) {
+          var <- as.matrix(var)
+          rownames(var) <- private$param_names$varied
+        }
+
+        colnames(var) <- self$populations()
+        t(var)
+      }
+      if (type == "fixed") {
+        pfix(theta, scale)
+      } else if (type == "varied") {
+        pvar(theta, scale)
+      } else {
+        fix <- pfix(theta, scale)
+        var <- pvar(theta, scale)
+        if (is.null(fix)) {
+          var
+        } else {
+          fix[, private$param_names$varied] <- var[, private$param_names$varied]
+          fix
+        }
+      }
+    },
+
+    ##' @description Apply the model transformation function to a parameter
+    ##' vector.
+    ##'
+    ##' @param theta a parameter vector in the same order as your
+    ##' parameters were defined in (see `$names()` for that order.
+    model = function(theta) {
+      apply(theta, 1, private$transform)
```

Comment:
I am not convinced that this will *always* return a list unless `private$transform` does. I think that's fine for now because dust requires a list for both

Reply:
That was my logic too, i.e. as dust requires it then this should always be the case